SO-4992: ETag and Cache-Control header support

cmark opened the pull request:
This PR adds support for `ETag` and `Cache-Control` headers so that HTTP clients, proxies, and intermediaries can cache the responses earlier than caching only in an HTTP browser. 

~Still somewhat blocked by https://github.com/spring-projects/spring-security/issues/12865. We need to figure out a way of preventing Spring from adding the default cache-control headers when we set our own.~

Resolved the above in the meantime by explicitly setting the headers in the response before Spring can write its defaults incorrectly.

Background: https://snowowl.atlassian.net/browse/SO-4992

cmark committed:
refactor(core): move rate limit configuration to RateLimitConfig class

cmark committed:
feat(core): support setting Cache-Control and ETag headers

cmark committed:
feat(api): propagate If-None-Match header to core system

cmark committed:
feat(api): handle NotModifiedException as HTTP 304 Not Modified response

cmark committed:
test: add basic Cache-Control and ETag header test cases

cmark committed:
chore(api): resolve compile errors in RateLimitConfig after cherry-pick

cmark committed:
feat(index): move ETag value computation to RevisionBranchRef

This provides the most accurate ETag value that can be used for any kind
of content request, be it versioned or not.

cmark committed:
chore: fix license text in cherry-picked files

cmark committed:
fix(api): write custom headers immediately into the HTTP response...

...when we build the response upon receiving a resolved/rejected
promise. This prevents header duplication and hopefully overrides
anything that Spring incorrectly adds by default.

cmark committed:
test(api): resolve SnomedApiCacheControlTest issues

Add it to AllSnomedApiTests suite.

cmark committed:
test: resolve failure in SnomedPartialLoadingApiTest

apeteri reviewed (approved):
👍

cmark committed:
test(api): add three more test cases to verify ETag and IfNoneMatch...

...header handling

cmark committed:
chore(test): use existing test helper to assert Location header presence

cmark committed:
chore(api): use RestController annotation instead of simple Controller

cmark committed:
fix(api): merge headers coming from promise response and from...

...promise body if the body is a Spring ResponseEntity instance

cmark committed:
chore(api): change since tag to 9.2

nagyo commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/rate/RateLimitConfig.java:
```diff
@@ -0,0 +1,47 @@
+public class RateLimitConfig {
+
+	@Min(0)
+	private long overdraft = 0L;
```
I would love to see a simple one-sentence explanation about the purpose of these values. For future devs I think it would be extremely useful.

cmark replied:
Done

nagyo reviewed (approved):
RSLGTM!

cmark committed:
chore(deps): upgrade to bucket4j 8.10.1

cmark committed:
feat(api): introduce `api.rate_limit.capacity` configuration...

...which replaces the `api.rate_limit.overdraft` config key

Add some documentation references to ease understanding of how
rate-limiting in Snow Owl works.

apeteri reviewed (approved):
👍

nagyo commented:
👌🏻
